lib/src/gotrue_client.dart

```diff
@@ -255,6 +255,13 @@ class GoTrueClient {
     if (response.error != null) return response;
 
     currentUser = response.user;
+    currentSession = Session(
+        accessToken: currentSession.accessToken,
+        expiresIn: currentSession.expiresAt,
+        refreshToken: currentSession.refreshToken,
+        tokenType: currentSession.tokenType,
+        providerToken: currentSession.providerToken,
+        user: response.user);
```

Comment:
I think we should create a `copyWith` method instead of this, because whenever we add a new property to `Session`, we'd need to add it here too.

Reply:
I have created a copyWith method as part of this PR: https://github.com/supabase/gotrue-dart/pull/37

Feel free to cherry pick

Reply:
Yes, I agree we should have a `copyWith()` method. I will update the PR

Reply:
One problem with @point-source implementation of `copyWith()` is that it is not going to work if `user` (or any other parameter) is null. We will have to think of a different approach

```dart
Session copyWith({
    String? accessToken,
    int? expiresIn,
    String? refreshToken,
    String? tokenType,
    String? providerToken,
    User? user,
  }) {
    return Session(
      accessToken: accessToken ?? this.accessToken,
      expiresIn: expiresIn ?? this.expiresIn,
      refreshToken: refreshToken ?? this.refreshToken,
      tokenType: tokenType ?? this.tokenType,
      providerToken: providerToken ?? this.providerToken,
      user: user ?? this.user,
    );
  }
```

Reply:
@bentesha wdym? It looks ok to me

Reply:
@bdlukaa Because the `Session.user` (and other fileds) is a nullable type. Consider the following example:
```dart
final someUser = getUserSomeHow();
final session = Session(accessToken: 'token', user: someUser);
final sessionCopy = session.copyWith(user: null);
assert(sessionCopy.user == null); // This will not pass!
assert(sessionCopy.user == someUser); // This will pass, which is not what is intended
```

Reply:
but if you want to remove the user, you'll also want to remove all the other fields, no? In that situation, just apply session to an empty Session

Reply:
@bdlukaa If a method does not work according to what it says, that is a buggy method. The method should create a copy of an object with new values, regardless of what values you pass to it. It should work consistently with all valid values.

Reply:
@bentesha I agree with you on this. We should probably not allow the user to be null and instead use a default instance of User which is empty or has an anonymous / not signed in flag. Unfortunately, this would be a somewhat significant divergence from the gotrue-js library so I'm not sure if this is the path the maintainers would like to take.

What other ways could we solve this?

Reply:
still... "if you want to remove the user, you'll also want to remove all the other fields, no? In that situation, just apply session to an empty Session" ... is the best option. I see nothing wrong on the current copyWith method. If you want to remove the current user, just set the session itself to null

Reply:
I agree with @bdlukaa and do not see a problem with the current implementation of `copyWith`. Although what you said about how `null` values cannot be set with `copyWith` method is true and inconvenient, this is how `copyWith` is implemented this way across many Flutter projects including [Flutter itself](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/theme_data.dart#L1390).